### ISSUE TITLE
Entropy metric may be incorrect when using VERL<=0.4.0

### ISSUE THREAD

**Issue by lee-junjie** [thunlp/JustRL#5]
When running a few training steps with the latest VERL (0.7.0), I observed that the entropy starts at approximately ~1.0. However, when using the code you previously shared that based on VERL 0.4.0, the entropy starts at approximately ~0.6.

(I shared my early observation here: https://github.com/thunlp/JustRL/issues/3#issuecomment-3705573747)

After further investigation, I found that this discrepancy appears to be caused by a bug in the entropy metric when `use_dynamic_bsz == True` in VERL 0.4.0. 

Specifically, in https://github.com/thunlp/JustRL/blob/main/train/verl/workers/actor/dp_actor.py#L309

The original implementation can cause a mismatch between the entropy tensor and the corresponding masking:

```python
if use_dynamic_bsz:
            indices = list(itertools.chain.from_iterable(indices))
            assert len(indices) == log_probs.size(0), f"{len(indices)} vs. {log_probs.size()}"
            revert_indices = torch.tensor(get_reverse_idx(indices), dtype=torch.long)
            log_probs = log_probs[revert_indices]
```
The entropy tensor should also be reordered consistently. A corrected version would be:
```python
if use_dynamic_bsz:
            indices = list(itertools.chain.from_iterable(indices))
            assert len(indices) == log_probs.size(0), f"{len(indices)} vs. {log_probs.size()}"
            revert_indices = torch.tensor(get_reverse_idx(indices), dtype=torch.long)
            log_probs = log_probs[revert_indices]
            if calculate_entropy and entropys is not None:
                entropys = entropys[revert_indices]
```

Note that VERL 0.4.0 was released on 2025-06-07, and this issue was fixed on 2025-06-15. You can see the corresponding fix here:
https://github.com/verl-project/verl/pull/1943/changes#diff-e06e73d3a7775a502b7aea91103e7911f6597eb48e4b898db558766cdd41daf9

After applying the fix, you should be able to see the entropy starts at ~1.0, or > 0.9 depends on how you process the prompt (whether to keep original prompt template).

Thanks again for your great work and look forward to your updated results!

**Comment by HBX-hbx** [comment]
Thanks for sharing this! In Figure 2(a) of the paper, the entropy indeed starts at ~0.6 and we have turned on use_dynamic_bsz. Not sure whether this bug will cause any discrepancy in the entropy during the later stage of training. Thank you again for pointing this out.

**Comment by hljjjmssyh** [comment]
Thanks for the insightful discussion above regarding the entropy bug in earlier versions. I have a follow-up question regarding the entropy behavior:

@lee-junjie  Have you successfully reproduced the results from the paper/blog after applying the fix or switching to the latest version of verl?

@HBX-hbx  We are currently using the latest verl version with the provided recipe, but we’ve observed that the entropy gradually decreases as training progresses. This trend seems quite different from the stable (or even slightly increasing) entropy behavior shown in Figure 2 of the paper.

I would appreciate your guidance on a few points:

Have you tried running the current recipe on recent versions of verl? 
Regarding the "gradually decreasing entropy" we observed, do you have any suggestions on how to maintain the entropy level as seen in your paper? For instance, are there specific hyperparameters (like kl_coeff, entropy_coeff, or settings related to dynamic_bsz) that might need adjustment in the newer verl framework to match your original results?
Looking forward to your thoughts. Thanks again!

**Comment by HBX-hbx** [comment]
> Thanks for the insightful discussion above regarding the entropy bug in earlier versions. I have a follow-up question regarding the entropy behavior:
> 
> [@lee-junjie](https://github.com/lee-junjie) Have you successfully reproduced the results from the paper/blog after applying the fix or switching to the latest version of verl?
> 
> [@HBX-hbx](https://github.com/HBX-hbx) We are currently using the latest verl version with the provided recipe, but we’ve observed that the entropy gradually decreases as training progresses. This trend seems quite different from the stable (or even slightly increasing) entropy behavior shown in Figure 2 of the paper.
> 
> I would appreciate your guidance on a few points:
> 
> Have you tried running the current recipe on recent versions of verl? Regarding the "gradually decreasing entropy" we observed, do you have any suggestions on how to maintain the entropy level as seen in your paper? For instance, are there specific hyperparameters (like kl_coeff, entropy_coeff, or settings related to dynamic_bsz) that might need adjustment in the newer verl framework to match your original results? Looking forward to your thoughts. Thanks again!

Thanks for the question. We haven't run experiments on the latest verl yet. In our experience, different inference engine versions can affect the output distribution and thus entropy. For large‑scale RL, we usually use a clip ratio (currently 0.28; if the entropy still decreases, try a higher ratio) and don't apply KL/entropy regularization (kl_coeff=0, entropy_coeff=0). That may explain the difference.

**Comment by lee-junjie** [comment]
> Thanks for the insightful discussion above regarding the entropy bug in earlier versions. I have a follow-up question regarding the entropy behavior:
> 
> [@lee-junjie](https://github.com/lee-junjie) Have you successfully reproduced the results from the paper/blog after applying the fix or switching to the latest version of verl?
> 
> [@HBX-hbx](https://github.com/HBX-hbx) We are currently using the latest verl version with the provided recipe, but we’ve observed that the entropy gradually decreases as training progresses. This trend seems quite different from the stable (or even slightly increasing) entropy behavior shown in Figure 2 of the paper.
> 
> I would appreciate your guidance on a few points:
> 
> Have you tried running the current recipe on recent versions of verl? Regarding the "gradually decreasing entropy" we observed, do you have any suggestions on how to maintain the entropy level as seen in your paper? For instance, are there specific hyperparameters (like kl_coeff, entropy_coeff, or settings related to dynamic_bsz) that might need adjustment in the newer verl framework to match your original results? Looking forward to your thoughts. Thanks again!

@hljjjmssyh Thanks for the question. I have run the repo using both the latest version and the Docker image of verl 0.4, and both show a similar trend of decreasing entropy after training. I also ran the codebase three times using the verl 0.4 image on H-series GPUs, but I was unable to reproduce the reported result. I’m not sure whether this discrepancy is caused by differences in GPU hardware or another factor.
